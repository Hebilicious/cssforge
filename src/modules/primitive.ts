import type { ColorConfig } from "./colors.ts";
import { processColors } from "./colors.ts";
import type { TypographyConfig } from "./typography.ts";
import { processTypography } from "./typography.ts";
import type { PixelSettings, SpacingConfig } from "./spacing.ts";
import { processSpacing } from "./spacing.ts";
import {
  getResolvedVariablesMap,
  type Output,
  resolveValue,
  type Variables,
} from "../lib.ts";
import { validateName } from "../helpers.ts";
import { pxToRem } from "../helpers.ts";

interface PrimitiveProperties {
  [key: string]: string;
}

interface PrimitiveDefinition {
  value: PrimitiveProperties;
  variables?: Variables;
  settings?: PixelSettings;
}

interface PrimitiveValue {
  [key: string]: PrimitiveDefinition;
}

/**
 * A single primitive definition.
 */
export interface Primitive {
  value: PrimitiveValue;
  settings?: unknown;
}

/**
 * A collection of primitives.
 */
export interface PrimitiveConfig {
  [key: string]: Primitive;
}

/**
 * Processes the primitive configuration to generate CSS variables.
 * Primitives can reference other design tokens from colors, typography, and spacing.
 * @example
 * ```ts
 * const config = {
 *  primitives: {
 *    button: {
 *      value: {
 *        default: {
 *          value: {
 *            "background-color": "var(--bg)",
 *          },
 *          variables: {
 *            "bg": "colors.palette.value.red.100",
 *          },
 *        },
 *      },
 *    },
 *  },
 *  colors: {
 *    palette: {
 *      value: {
 *        red: {
 *          100: { hex: "#ff0000" },
 *        },
 *      },
 *    },
 *  },
 * };
 * const output = processPrimitives(config);
 * // output.css: "/* button * / ;\n--button-default-background-color: var(--color-red-100);"
 * ```
 */
export function processPrimitives(
  config: {
    primitives: PrimitiveConfig;
    colors?: ColorConfig;
    typography?: TypographyConfig;
    spacing?: SpacingConfig;
  },
): Output {
  const cssOutput: string[] = [];
  const resolveMap = new Map();
  const moduleKey = "primitives";
  const colorVariables = config.colors ? processColors(config.colors) : null;
  const typographyVariables = config.typography
    ? processTypography(config.typography)
    : null;
  const spacingVariables = config.spacing ? processSpacing(config.spacing) : null;
  for (const [primitiveName, primitive] of Object.entries(config.primitives)) {
    validateName(primitiveName);
    cssOutput.push(`/* ${primitiveName} */`);
    for (
      const [
        variantName,
        { value: properties, variables, settings = { pxToRem: true, rem: 16 } },
      ] of Object.entries(
        primitive.value,
      )
    ) {
      validateName(variantName);
      try {
        const resolvedMap = getResolvedVariablesMap({
          variables,
          colors: colorVariables,
          typography: typographyVariables,
          spacing: spacingVariables,
        });

        for (const [propName, propValue] of Object.entries(properties)) {
          validateName(propName);

          const convertedValue = settings.pxToRem
            ? pxToRem({ value: propValue, rem: settings.rem })
            : propValue;

          const resolvedValue = resolveValue({ map: resolvedMap, value: convertedValue });

          const key = `--${primitiveName}-${variantName}-${propName}`;
          const variable = `${key}: ${resolvedValue};`;
          cssOutput.push(variable);
          resolveMap.set(
            `${moduleKey}.${primitiveName}.value.${variantName}.${propName}`,
            {
              key,
              value: resolvedValue,
              variable,
            },
          );
        }
      } catch (error) {
        console.error(
          `Error processing primitive ${primitiveName}-${variantName}:`,
          error,
        );
        throw error;
      }
    }
  }

  return { css: cssOutput.join("\n"), resolveMap };
}
