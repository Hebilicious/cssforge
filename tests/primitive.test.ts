import { assertEquals } from "@std/assert";
import { processPrimitives } from "../src/modules/primitive.ts";
import { defineConfig } from "../src/mod.ts";

Deno.test("processPrimitives - processes button with variables", () => {
  const config = defineConfig({
    typography: {
      fluid: {
        arial: {
          value: {
            minWidth: 320,
            minFontSize: 14,
            minTypeScale: 1.25,
            maxWidth: 1435,
            maxFontSize: 16,
            maxTypeScale: 1.25,
            positiveSteps: 5,
            negativeSteps: 3,
            prefix: "foo",
          },
        },
      },
    },
    spacing: {
      custom: {
        size: {
          value: {
            2: "0.5rem",
            3: "0.75rem",
          },
        },
      },
    },
    primitives: {
      button: {
        value: {
          small: {
            value: {
              width: "120px",
              height: "40px",
              fontSize: "var(--base)",
              radius: "8px",
              padding: "var(--2) var(--3)",
            },
            variables: {
              "base": "typography_fluid.arial@m",
              "2": "spacing.custom.size.value.2",
              "3": "spacing.custom.size.value.3",
            },
            settings: { pxToRem: false },
          },
        },
      },
    },
  });

  const result = processPrimitives(config);
  const expected = [
    "/* button */",
    "--button-small-width: 120px;",
    "--button-small-height: 40px;",
    "--button-small-fontSize: var(--typography_fluid-arial-foo-m);",
    "--button-small-radius: 8px;",
    "--button-small-padding: var(--spacing-size-2) var(--spacing-size-3);",
  ].join("\n");

  assertEquals(result.css, expected);
});

Deno.test("processPrimitives - processes buttons with settings", () => {
  const config = defineConfig({
    primitives: {
      button: {
        value: {
          small: {
            value: {
              width: "120px",
              height: "40px",
              radius: "8px",
            },
            settings: { pxToRem: false },
          },
          big: {
            value: {
              width: "240px",
              height: "80px",
              radius: "16px",
            },
            settings: { pxToRem: true },
          },
        },
      },
    },
  });

  const result = processPrimitives(config);
  const expected = [
    "/* button */",
    "--button-small-width: 120px;",
    "--button-small-height: 40px;",
    "--button-small-radius: 8px;",
    "--button-big-width: 15rem;",
    "--button-big-height: 5rem;",
    "--button-big-radius: 1rem;",
  ].join("\n");

  assertEquals(result.css, expected);
});

Deno.test("processPrimitives - references colors, gradients, and themes", () => {
  const config = defineConfig({
    colors: {
      palette: {
        value: {
          coral: {
            "50": { hex: "#FF7E60" },
          },
        },
      },
      gradients: {
        value: {
          orangeGradient: {
            value: {
              primary: {
                value: "linear-gradient(to right, var(--c1), var(--c2))",
                variables: {
                  "c1": "palette.value.coral.50",
                  "c2": "palette.value.coral.50",
                },
              },
            },
          },
        },
      },
      theme: {
        value: {
          light: {
            background: {
              value: {
                primary: "var(--grad)",
              },
              variables: {
                "grad": "gradients.value.orangeGradient.primary",
              },
            },
          },
        },
      },
    },
    primitives: {
      card: {
        value: {
          default: {
            value: {
              "background-color": "var(--bg)",
              "background-image": "var(--grad)",
              "border-color": "var(--border)",
            },
            variables: {
              "bg": "theme.value.light.background.primary",
              "grad": "gradients.value.orangeGradient.primary",
              "border": "palette.value.coral.50",
            },
          },
        },
      },
    },
  });

  const result = processPrimitives(config);
  const expected = [
    "/* card */",
    "--card-default-background-color: var(--theme-light-background-primary);",
    "--card-default-background-image: var(--gradients-orangeGradient-primary);",
    "--card-default-border-color: var(--palette-coral-50);",
  ].join("\n");

  assertEquals(result.css, expected);
});

Deno.test("processPrimitives - uses fluid spacing references", () => {
  const config = defineConfig({
    spacing: {
      fluid: {
        gap: {
          value: {
            minSize: 4,
            maxSize: 24,
            minWidth: 320,
            maxWidth: 1280,
            negativeSteps: [0],
            positiveSteps: [1],
            prefix: "gs",
          },
        },
      },
    },
    primitives: {
      box: {
        value: {
          default: {
            value: {
              padding: "var(--s) var(--m)",
            },
            variables: {
              "s": "spacing_fluid.gap@s",
              "m": "spacing_fluid.gap@m",
            },
            settings: { pxToRem: false },
          },
        },
      },
    },
  });
  const primitives = processPrimitives(config);
  const expected = [
    "/* box */",
    "--box-default-padding: var(--spacing_fluid-gap-gs-s) var(--spacing_fluid-gap-gs-m);",
  ].join("\n");
  assertEquals(primitives.css, expected);
});
