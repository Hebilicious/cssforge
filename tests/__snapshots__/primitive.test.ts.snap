export const snapshot = {};

snapshot[`processPrimitives - processes button with variables 1`] = `
"/* button */
--button-small-width: 120px;
--button-small-height: 40px;
--button-small-fontSize: var(--typography_fluid-arial-foo-m);
--button-small-radius: 8px;
--button-small-padding: var(--spacing-size-2) var(--spacing-size-3);"
`;

snapshot[`processPrimitives - processes button with variables 2`] = `
[
  [
    "primitives.button.small.width",
    {
      key: "--button-small-width",
      value: "120px",
      variable: "--button-small-width: 120px;",
    },
  ],
  [
    "primitives.button.small.height",
    {
      key: "--button-small-height",
      value: "40px",
      variable: "--button-small-height: 40px;",
    },
  ],
  [
    "primitives.button.small.fontSize",
    {
      key: "--button-small-fontSize",
      value: "var(--typography_fluid-arial-foo-m)",
      variable: "--button-small-fontSize: var(--typography_fluid-arial-foo-m);",
    },
  ],
  [
    "primitives.button.small.radius",
    {
      key: "--button-small-radius",
      value: "8px",
      variable: "--button-small-radius: 8px;",
    },
  ],
  [
    "primitives.button.small.padding",
    {
      key: "--button-small-padding",
      value: "var(--spacing-size-2) var(--spacing-size-3)",
      variable: "--button-small-padding: var(--spacing-size-2) var(--spacing-size-3);",
    },
  ],
]
`;

snapshot[`processPrimitives - processes buttons with settings 1`] = `
"/* button */
--button-small-width: 120px;
--button-small-height: 40px;
--button-small-radius: 8px;
--button-big-width: 15rem;
--button-big-height: 5rem;
--button-big-radius: 1rem;"
`;

snapshot[`processPrimitives - processes buttons with settings 2`] = `
[
  [
    "primitives.button.small.width",
    {
      key: "--button-small-width",
      value: "120px",
      variable: "--button-small-width: 120px;",
    },
  ],
  [
    "primitives.button.small.height",
    {
      key: "--button-small-height",
      value: "40px",
      variable: "--button-small-height: 40px;",
    },
  ],
  [
    "primitives.button.small.radius",
    {
      key: "--button-small-radius",
      value: "8px",
      variable: "--button-small-radius: 8px;",
    },
  ],
  [
    "primitives.button.big.width",
    {
      key: "--button-big-width",
      value: "15rem",
      variable: "--button-big-width: 15rem;",
    },
  ],
  [
    "primitives.button.big.height",
    {
      key: "--button-big-height",
      value: "5rem",
      variable: "--button-big-height: 5rem;",
    },
  ],
  [
    "primitives.button.big.radius",
    {
      key: "--button-big-radius",
      value: "1rem",
      variable: "--button-big-radius: 1rem;",
    },
  ],
]
`;

snapshot[`processPrimitives - references colors, gradients, and themes 1`] = `
"/* card */
--card-default-background-color: var(--theme-light-background-primary);
--card-default-background-image: var(--gradients-orangeGradient-primary);
--card-default-border-color: var(--palette-coral-50);"
`;

snapshot[`processPrimitives - references colors, gradients, and themes 2`] = `
[
  [
    "primitives.card.default.background-color",
    {
      key: "--card-default-background-color",
      value: "var(--theme-light-background-primary)",
      variable: "--card-default-background-color: var(--theme-light-background-primary);",
    },
  ],
  [
    "primitives.card.default.background-image",
    {
      key: "--card-default-background-image",
      value: "var(--gradients-orangeGradient-primary)",
      variable: "--card-default-background-image: var(--gradients-orangeGradient-primary);",
    },
  ],
  [
    "primitives.card.default.border-color",
    {
      key: "--card-default-border-color",
      value: "var(--palette-coral-50)",
      variable: "--card-default-border-color: var(--palette-coral-50);",
    },
  ],
]
`;

snapshot[`processPrimitives - uses fluid spacing references 1`] = `
"/* box */
--box-default-padding: var(--spacing_fluid-gap-gs-s) var(--spacing_fluid-gap-gs-m);"
`;

snapshot[`processPrimitives - uses fluid spacing references 2`] = `
[
  [
    "primitives.box.default.padding",
    {
      key: "--box-default-padding",
      value: "var(--spacing_fluid-gap-gs-s) var(--spacing_fluid-gap-gs-m)",
      variable: "--box-default-padding: var(--spacing_fluid-gap-gs-s) var(--spacing_fluid-gap-gs-m);",
    },
  ],
]
`;
